core/drivers/regulator/regulator_dt.c

```diff
@@ -0,0 +1,368 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022-2023, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <drivers/regulator.h>
+#include <initcall.h>
+#include <kernel/dt_driver.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <util.h>
+
+/*
+ * struct pending_regu - Regulators waiting for their supply to be ready
+ *
+ * @fdt: DT to work on
+ * @node: Node of the regulator in @fdt
+ * @supply_phandle: Phandle in @fdt of the regulator supply, or 0 if no supply
+ * @regulator_allocated: True if framework allocates and frees @regulator
+ * @regulator: Regulator device instance
+ * @link: Link in pending regulators list
+ *
+ * When calling regulator_dt_register(), either the regulator depends on a
+ * supply that is not initialized, or this dependency is resolved (there is
+ * no supply or the supply is ready to use).
+ *
+ * In the former case, the regulator is placed in a pending regulator list.
+ * Each time a new regulator is successfully registered, we process the
+ * pending regulator list in case some pending regulators find their
+ * supply and finalize their registration and initialization.
+ *
+ * In the latter case, the regulator registration and initialization
+ * are processed.
+ */
+struct pending_regu {
+	const void *fdt;
+	int node;
+	int supply_phandle;
+	bool regulator_allocated;
+	struct regulator *regulator;
+	SLIST_ENTRY(pending_regu) link;
+};
+
+static SLIST_HEAD(, pending_regu) pending_regu_list =
+	SLIST_HEAD_INITIALIZER(pending_regu);
+
+/* Helper to find the phandle of a regulator supply */
+static TEE_Result get_supply_phandle(const void *fdt, int node,
+				     const char *supply_name,
+				     uint32_t *supply_phandle)
+{
+	char supply_prop_stack[40] = { };
+	char *supply_prop = NULL;
+	size_t prop_len = 0;
+	const fdt32_t *cuint = NULL;
+	int len = 0;
+
+	prop_len = strlen(supply_name) + strlen("-supply") + 1;
+
+	if (prop_len > sizeof(supply_prop_stack)) {
+		supply_prop = calloc(1, prop_len);
```

Comment:
Why not always use `calloc()`? Is there something noticeable to be gained with this code?

Reply:
no strong reasons. i think i've had execution speed concerns time. There is no heap fragmentation concerns here.
You think it's not worth it ?

Reply:
I bet this is only marginally faster (I have not tried any profiling though), so I'd rather keep the code as simple as possible.